js/src/data/actions.js

```diff
@@ -72,77 +73,80 @@ export function* fetchShippingRates() {
 }
 
 /**
- * Aggregated shipping rate.
+ * Saves shipping rates.
  *
- * @typedef {Object} AggregatedShippingRate
- * @property {Array<CountryCode>} countries Array of destination country codes.
- * @property {string} currency Currency of the price.
- * @property {number} rate Shipping price.
- */
-
-/**
- * Updates or inserts given aggregated shipping rate.
+ * This is done by removing the old shipping rates first,
+ * and then upserting the new shipping rates.
  *
- * @param {AggregatedShippingRate} shippingRate
+ * @param {Array<ShippingRate>} newShippingRates
  */
-export function* upsertShippingRates( shippingRate ) {
-	const { countryCodes, currency, rate } = shippingRate;
-
+export function* saveShippingRates( newShippingRates ) {
 	try {
-		yield apiFetch( {
-			path: `${ API_NAMESPACE }/mc/shipping/rates/batch`,
-			method: 'POST',
-			data: {
-				country_codes: countryCodes,
-				currency,
-				rate,
-			},
-		} );
+		/**
+		 * Compare the new shipping rates and the old ones from the store
+		 * to find out the old ones to be deleted.
+		 */
+		const oldShippingRates = yield select( STORE_KEY, 'getShippingRates' );
```

Comment:
❔ Would you consider passing `oldShippingRates` through the function caller instead of getting it inside this function? The reason is basically the same as https://github.com/woocommerce/google-listings-and-ads/pull/1262#discussion_r814721197.

Reply:
It is more flexible to design action with as much closer single responsibility as possible since actions are the elementary data control interfaces on frontend side.

For example, the current design of `saveShippingRates` is fully overwritten all rates, which is like a RESTful **PUT** API call. When one or more shipping rates need to be deleted, the current state must be obtained from wp-data store, and filter out the rates to be deleted before calling `saveShippingRates`; when some shipping rates need to be updated, the current state still must be obtained from wp-data store, use that state to update the required ones, and then call `saveShippingRates` along with all rates that include those ones don't need updates.

That is, except it looks much cleaner on the caller side when clearing all shipping rates. But in other places, the caller always already has the rates state obtained from wp-data store. And with calling of useShippingRates, clearing all rates doesn't add much complexity.

Moreover, they also cause redundant API calls and database operations on the server-side internally. Especially when many countries are selected, the auto-save feature causes more redundant calls and loads to server.
https://github.com/woocommerce/google-listings-and-ads/blob/0cca8d9e92d620d21688dedf3bc2053c6f70cb07/src/API/Site/Controllers/MerchantCenter/ShippingRateBatchController.php#L54-L58

Moving `oldShippingRates` out of `saveShippingRates` could preserve a certain degree of flexibility for easier detail optimization in the future. And it should be easier to write tests for `saveShippingRates` because it would not need to mock `select`.

---

If go further into abstraction design, I would prefer to keep the `upsertShippingRates` and `deleteShippingRates` actions as they were exported originally. The `saveShippingRates` would be better suited as a custom hook, since it actually handles business logics that are closer to UI-wise, but now those business logics are highly coupled into data layer with data processing logic. In this way,
- the caller side could still not need to obtain the current shipping rates state if caller doesn't have it.
- the data layer would have more flexibility.
- it's more stateless so that writing tests should be easier.
- the interfaces of `upsert*` and `delete*` actions would be more clear than `saveShippingRates( oldShippingRates, newShippingRates )` when partially updating/deleting.

Reply:
> I would prefer to keep the `upsertShippingRates` and `deleteShippingRates` actions as they were exported originally. _The `saveShippingRates` would be better suited as a custom hook_, since it actually handles business logics that are closer to UI-wise

Hey I think that is a good idea! 😃 👍  So to summarize, I think this is what we are thinking:

1. In wp.data, we don't need `saveShippingRates`, we want `upsertShippingRates` and `deleteShippingRates` actions.
2. We come up with a custom hook like the following, which consumes the above `upsertShippingRates` and `deleteShippingRates` actions:

```js
// an example:
const useSaveShippingRate = () => {
	const saveShippingRate = useCallback( () => {
		// do the deleteShippingRates and upsertShippingRates here.
	}, [])

	return saveShippingRate
}
```

3. Change the code in Edit Free Listings and Setup MC to use the above hook.

This approach also addresses @jorgemd24's comment in https://github.com/woocommerce/google-listings-and-ads/pull/1262#discussion_r814721197, to have better function composition.

Given that the current `saveShippingRate` action is being used in Edit Free Listings PR https://github.com/woocommerce/google-listings-and-ads/pull/1266 and Setup MC PR https://github.com/woocommerce/google-listings-and-ads/pull/1267, I will work on that change as a separate PR later based on https://github.com/woocommerce/google-listings-and-ads/pull/1267.

Reply:
> At first I thought about that (before I found out about `yield select` in actions)

Side note for posterity: I found put about the `yield select` in actions from here: https://unfoldingneurons.com/2020/wordpress-data-store-properties-action-creator-generators

Reply:
> 1\. [...]
> 2\. [...]
> 3\. Change the code in Edit Free Listings and Setup MC to use the above hook.

Yes, the `useSaveShippingRate` could still do the same thing inside the hook, and further have the improvement of https://github.com/woocommerce/google-listings-and-ads/pull/1262#discussion_r817858184. And if some other callers already know what to upsert/delete, they could call `upsertShippingRates`/`deleteShippingRates` directly.

> Given that the current `saveShippingRate` action is being used in Edit Free Listings PR #1266 and Setup MC PR #1267, I will work on that change as a separate PR later based on #1267.

> With this approach, we will only call upsertShippingRate for those shipping rates that are newly created or have been changed, and the consumer would still call the same API `saveShippingRates (newShippingRates)`, which is nice.

Agree with these. 👍  Thanks for considering these options!

Reply:
> Given that the current `saveShippingRate` action is being used in Edit Free Listings PR #1266 and Setup MC PR #1267, I will work on that change as a separate PR later based on #1267.

FYI, I have created issue https://github.com/woocommerce/google-listings-and-ads/issues/1287 to keep track of this.